### PR TITLE
[InstCombine] Improve coverage of `foldSelectValueEquivalence` for non-constants

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[InstCombine] Add tests for expanding `foldSelectValueEquivalence`; NFC**
- **[InstCombine] Improve coverage of `foldSelectValueEquivalence` for non-constants**

There are a few more cases we can easily cover.
   - If f(Y) simplifies to Y
   - If replace X with Y makes X one-use (and make Y non one-use).

These all require that Y is not undef

### CODE REVIEW

**Comment by goldsteinn** [comment]
Going to run csmith for a few days w/ this. Will ping back results probably monday.

**Review by nikic — CHANGES_REQUESTED** [review]
I don't think this is worth the complication. The only part of this I would consider is the `NewOp == V` case, though I'm having some trouble convincing myself that this really can't lead to a combine loop.

**Comment by goldsteinn** [comment]
Ran ~5million csmith tests. No inf loops or bad codegen. Although note only got different resulting IR ~100 times over all the tests.

**Comment by dtcxzyw** [comment]
> I don't think this is worth the complication. The only part of this I would consider is the `NewOp == V` case, though I'm having some trouble convincing myself that this really can't lead to a combine loop.

I agree.

**Comment by goldsteinn** [comment]
Dropped the use-reduction code. Think for this to inf loop you would need to violate SSA i.e: f(x, y) -> y -> y(x, z) -> x. For us to go from x -> y, then y would need to both be dominated by x and dominate x.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 1]
```diff
@@ -1308,13 +1310,20 @@ Instruction *InstCombinerImpl::foldSelectValueEquivalence(SelectInst &Sel,
           isGuaranteedNotToBeUndef(V, SQ.AC, &Sel, &DT))
         return replaceOperand(Sel, Swapped ? 2 : 1, V);
 
-      // If NewOp is a constant and OldOp is not replace iff NewOp doesn't
-      // contain and undef elements.
-      if (match(NewOp, m_ImmConstant())) {
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:1313]: Can you just add `|| NewOp == V` to this line and leave everything else alone?

**Review by nikic — APPROVED** [review]
LGTM

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Pull request closed without merging**